### PR TITLE
Compilation fixes

### PR DESCRIPTION
**Pull request by nicholasbishop**
`cargo test` succeeds for me with these changes, tested with `rustc 1.0.0-nightly (d17d6e7f1 2015-04-02) (built 2015-04-03)`

### CODE REVIEW

File: src/render/batch.rs [hunk 1]
```diff
@@ -160,9 +160,10 @@ impl<T: ShaderParam> Batch for OwnedBatch<T> {
 type Index = u16;
 
 //#[derive(PartialEq, Eq, PartialOrd, Ord, Debug)]
+#[derive(Clone)]
 struct Id<T>(Index, PhantomData<T>);
 
-impl<T> Copy for Id<T> {}
+impl<T: Clone> Copy for Id<T> {}
```
Comment (bjz) [inline src/render/batch.rs]: Could you instead do:
~~~rust
impl<T> Clone for Id<T> {
    fn clone(&self) -> Id<T> { *self }
}
~~~

File: src/render/batch.rs [hunk 2]
```diff
@@ -252,7 +254,7 @@ pub struct CoreBatch<T: ShaderParam> {
     state_id: Id<DrawState>,
 }
 
-impl<T: ShaderParam> Copy for CoreBatch<T> where T::Link: Copy {}
+impl<T: Clone + ShaderParam> Copy for CoreBatch<T> where T::Link: Copy {}
```
Comment (bjz) [inline src/render/batch.rs]: I assume the need for the `Clone` bound would go away after my previous comment.

**Comment by bjz** [comment]
Thanks! A few comments. I would like to get this merged asap.

**Comment by bjz** [comment]
Actually, I'll merge now and make the changes after.

**Comment by nicholasbishop** [comment]
Thanks for the quick merge. Just for my own edification, could you explain a bit about why the explicit `Clone` impl is right way to go here? I'm still mostly at the stage of making blind edits when the compiler yells at me, so I'd be interested to hear a few words about the bigger picture.

**Comment by bjz** [comment]
No worries! Syntax extensions like `#[derive(..)]` have no knowledge of the types. So `derive` just blindly adds `Clone` bounds to all the type parameters. In our case, `Clone` does not need to be added to `T`, because it does not need to be cloned. Dunno if that makes sense...

**Comment by kvark** [comment]
Thanks @nicholasbishop !

**Comment by nicholasbishop** [comment]
Yep, that indeed makes sense. Thanks for the explanation :)

**Pull request merged**